transport/graphsync/dtchannel/dtchannel.go

```diff
@@ -262,6 +282,113 @@ func (c *Channel) Resume(ctx context.Context, extensions []graphsync.ExtensionDa
 	return c.gs.Unpause(ctx, *c.requestID, extensions...)
 }
 
+type Action string
+
+const (
+	NoAction Action = ""
+	Close    Action = "close"
+	Pause    Action = "pause"
+	Resume   Action = "resume"
+)
+
+// UpdateFromChannelState updates internal graphsync channel state form a datatransfer
+// channel state
+func (c *Channel) UpdateFromChannelState(chst datatransfer.ChannelState) error {
+	c.lk.Lock()
+	defer c.lk.Unlock()
+	return c.updateFromChannelState(chst)
+}
+
+func (c *Channel) updateFromChannelState(chst datatransfer.ChannelState) error {
+	// read the sent value
+	sentNode := chst.SentIndex()
+	if !sentNode.IsNull() {
```

Comment:
More of a comment than a change request: these `IsNull`s dictate the contract on `ChannelState` that is currently fulfilled by `CborGenCompatibleNode`. If we were to move away from cbor-gen then that contract would need to follow. By default the `Node`s present here would need to satisfy `IsNull()==true` in their default state—which isn't necessarily natural (e.g. maybe `Absent` or simply `nil` would be more natural choices).

:shrug: I think I'm just feeling the mild awkwardness of the implicit `datatransfer.ChannelState` contract being dictated by use this far down in the stack. Hopefully tests would pick up any problems if we changed later on without realising this.